lib/fastly_nsq/rake_task.rb

```diff
@@ -34,19 +34,21 @@ def add_rake_task_description_if_one_needed
     end
 
     def run_tasks
+      listener_threads = []
+
       topics.each do |topic|
-        Thread.new do
+        listener_threads << Thread.new do
           wrap_helpful_output(topic) do
             MessageQueue::Listener.new(topic: topic, channel: channel).go
           end
         end
       end
 
-      non_main_threads.map(&:join)
-    end
-
-    def non_main_threads
-      (Thread.list - [Thread.main])
+      while listener_threads.any?(&:status)
+        listener_threads.each do |thread|
+          thread.join(1)
```

Comment:
I had the same idea, but this will only loop each thread once. And should a thread raise before it becomes the subject of the loop it will be silently discarded. It also won't wait for threads to join because it will only wait one second and then return nil if the thread has not exited.

@alieander: I assume you used the `while` and the `join(1)` so that if one thread raised an error it will be caught here? What about using `abort_on_exception=true` so that when a thread raises we'll just raise and exit?